client/services/code_evaluators/PythonPrereqCheckService.js

```diff
@@ -0,0 +1,126 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for performing pre-requisite checks on Python code
+ * snippets.
+ */
+
+tie.factory('PythonPrereqCheckService', [
+  'CodePrereqCheckResultObjectFactory', 'PrereqCheckFailureObjectFactory',
+  'PREREQ_CHECK_TYPE_BAD_IMPORT', 'PREREQ_CHECK_TYPE_MISSING_STARTER_CODE',
+  'SUPPORTED_PYTHON_LIBS',
+  function(
+      CodePrereqCheckResultObjectFactory, PrereqCheckFailureObjectFactory,
+      PREREQ_CHECK_TYPE_BAD_IMPORT, PREREQ_CHECK_TYPE_MISSING_STARTER_CODE,
+      SUPPORTED_PYTHON_LIBS) {
+
+    var rightTrim = function(str) {
+      //remove trailing white space at end of string
+      var RIGHT_TRIM_PATTERN = /\s+$/;
+      return str.replace(RIGHT_TRIM_PATTERN, '');
+    };
+
+    var extractTopLevelFunctionLines = function(starterCode) {
+      var starterCodeLines = starterCode.split('\n');
+      var topLevelFunctionLines = [];
+      for (var i = 0; i < starterCodeLines.length; i++) {
+        var line = rightTrim(starterCodeLines[i]);
+        if (line.startsWith("def ")) {
+          topLevelFunctionLines.push(line);
+        }
+      }
+      return topLevelFunctionLines;
+    };
+
+    var doTopLevelFunctionLinesExist = function(
+      code, expectedTopLevelFunctionLines) {
+        var codeLines = code.split('\n');
+        for (var i = 0; i < codeLines.length; i++) {
+          codeLines[i] = rightTrim(codeLines[i]);
+        }
+
+        return expectedTopLevelFunctionLines.every(function(expectedLine) {
+          return codeLines.indexOf(expectedLine) !== -1;
+        });
+    };
+
+    var checkStarterCodeFunctionsPresent = function(starterCode, code) {
+      var expectedTopLevelFunctionLines = extractTopLevelFunctionLines(
+        starterCode);
+      return doTopLevelFunctionLinesExist(code, expectedTopLevelFunctionLines);
+    };
+
+    var getImportedLibraries = function(code) {
+      var codeLines = code.split('\n');
+      var importedLibraries = [];
+      var importPattern = new RegExp('^import\\ (\\w+)$');
+      for (var i = 0; i < codeLines.length; i++) {
+        var match =  importPattern.exec(codeLines[i]);
+        if (match) {
+          importedLibraries.push(match[1]);
+        }
+      }
+      return importedLibraries;
+    };
+
+    var getUnsupportedImports = function(importedLibraries) {
+      var unsupportedImports = importedLibraries.filter(function(
+        importedLibrary) {
+          return SUPPORTED_PYTHON_LIBS.indexOf(importedLibrary) === -1;
+        }
+      );
+      return unsupportedImports;
+    };
+
+    return {
+      // Returns a promise.
+      checkCode: function(starterCode, code) {
+        // check that starter code is present
```

Comment:
Same comments as above for the styling of all these comments (start with capital letter, end with period).